opendevin/sandbox/docker.py

```diff
@@ -0,0 +1,145 @@
+import os
+import pty
+import sys
+import uuid
+import time
+import shlex
+import select
+import subprocess
+from typing import List
+from collections import namedtuple
+
+InputType = namedtuple("InputDtype", ["content"])
+OutputType = namedtuple("OutputDtype", ["content"])
+
+
+class DockerInteractive:
+    CONTAINER_IMAGE = "opendevin/sandbox:latest"
+
+    def __init__(
+        self,
+        workspace_dir: str = None,
+        container_image: str = None,
+        timeout: int = 5
+    ):
+        self.instance_id: str = uuid.uuid4()
+        if workspace_dir is not None:
+            assert os.path.exists(workspace_dir), f"Directory {workspace_dir} does not exist."
+            # expand to absolute path
+            workspace_dir = os.path.abspath(workspace_dir)
+        else:
+            workspace_dir = os.getcwd()
+            print(f"workspace unspecified, using current directory: {workspace_dir}")
+        
+        # TODO: this timeout is actually essential - need a better way to set it
+        # if it is too short, the container may still waiting for previous
+        # command to finish (e.g. apt-get update)
+        # if it is too long, the user may have to wait for a unnecessary long time
+        self.timeout: int = timeout
+
+        if container_image is None:
+            container_image = self.CONTAINER_IMAGE
+
+        uid = os.getuid()
+        cmd = (
+            f"docker run -it --rm --name sandbox-{self.instance_id} "
+            f"-v {workspace_dir}:/workspace "
+            f"-w /workspace "
+            f"--network=host "
+            f"{container_image} "
+            f"/bin/bash -c 'useradd --shell /bin/bash -u {uid} -o -c \"\" -m devin && su devin'"
+        )
+        # print(f"Starting Docker container with command: {cmd}")
+        self.master_fd, self.slave_fd = pty.openpty()
+        self.container = subprocess.Popen(
+            shlex.split(cmd),
+            stdin=self.slave_fd,
+            stdout=self.slave_fd,
+            stderr=self.slave_fd,
+            text=True,
+            close_fds=True,
+        )
+        time.sleep(1)  # wait for the container to start
+        # TODO: use a more robust way to check if the container is ready
+        self.history: List[InputType | OutputType] = [
+            OutputType(self._wait_and_read_output())
+        ]
+
+    def _wait_and_read_output(self, user_input: str = None) -> str:
+        output_str = ""
+        while True:
+            readable, _, _ = select.select([self.master_fd], [], [], self.timeout)
+            if readable:
+                output = os.read(self.master_fd, 1024).decode()
+                if not output:
+                    break
+                output_str += output
+            else:
+                break
+        if user_input:
+            output_str = output_str.lstrip(user_input).lstrip()
+        return output_str
+
+    def execute(self, cmd: str) -> str:
```

Comment:
If I'm reading this right, to run multiple commands concurrently (e.g. `node server.js` and `curl localhost:3000`) we'll want to instantiate multiple sandboxes with different IDs. I think that will work well.